keps/sig-apps/3140-TimeZone-support-in-CronJob/README.md

```diff
@@ -0,0 +1,504 @@
+Zone used during the creation depens on where kube-controller-manager is running.
+This proposal aims to extend CronJob resource with the ability for a user to
+define the TimeZone when a Job should be created.
+
+## Motivation
+
+Not long after the [introduction of CronJob in kubernetes](https://github.com/kubernetes/kubernetes/pull/11980)
+a [request was raised to support setting time zones](https://github.com/kubernetes/kubernetes/issues/47202).
+The initial [response from SIG-Apps and SIG-Architecture](https://github.com/kubernetes/kubernetes/issues/47202#issuecomment-360820586)
+at the time was that introducing such functionality would require cluster operator
+to manually include TimeZone database since golang did not have one.
+Starting from [golang 1.15](https://go.dev/doc/go1.15) `time/tzdata` package can
+be embedded in the binary thus removing the requirement for external database.
+At that time the majority of the focus was towards [moving CronJob to GA](https://github.com/kubernetes/enhancements/issues/19)
+so the effort to support TimeZone was again delayed. Now that we have CronJob
+fully GA-ed it's time to satisfy the original request.
+
+### Goals
+
+- Add the field `.spec.timeZone` which allows specifying a valid TimeZone
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+## Proposal
+
+Add the field `.spec.timeZone` to the CronJob resource. The cronjob controller
+will take the field into account when scheduling the next Job run. In case the
+field is not specified or is empty the controller will maitain the current
+behavior which is to rely on the time zone of the kube-controller-manager
+process.
+
+### Notes/Constraints/Caveats (Optional)
+
+The current mechanism, which will still be the default behavior heavily relies
+on the time zone of the kube-controller-manager process which is hard for a
+regular user to figure out. Exposing an explicit field for setting a time zone
+will allow CronJob authors to simplify the user experience when creating CronJobs.
+
+### Risks and Mitigations
+
+- Outdated time zone in the golang might lead to wrong schedule times
+
+This problem can be mitigated with a fresh build of kube-controller-manager with
+updated golang version, but it heavily relies on the go community to keep the
+time zone database up-to-date.
+
+- Malicious user can create multiple CronJobs with different time zone which can
+actually trigger Jobs at the exact same time
+
+Cluster administrators should enforce quota to ensure not that many Jobs and
+CronJobs can be created per user.
+
+
+## Design Details
+
+### CronJob API
+
+The `CronJobSpec` structure is expanded with new `TimeZone` field which allows
+specifying the name of the time zone to be used. Missing or empty value of the
+field indicates the current behavior, which relies on the time zone of the
+kube-controller-manager process.
```

Comment:
```suggestion
The `.spec` for a CronJob is expanded with new `timeZone` field which allows
specifying the name of the time zone to be used. Missing or empty value of the
field indicates the current behavior, which relies on the time zone of the
kube-controller-manager process.

In the API code, that looks like:
```

1. Does the kube-controller-manager populate the `.spec.timeZone` once it detects a CronJob with no timezone set?
2. Should it?